src/main/java/com/epam/aidial/cfg/service/ToolSetResourceService.java

```diff
@@ -0,0 +1,144 @@
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.epam.aidial.cfg.client.mapper.ToolSetClientMapper.TOOLSETS_PREFIX;
+import static com.epam.aidial.cfg.utils.HeaderUtils.createHeadersForCreate;
+import static com.epam.aidial.cfg.utils.HeaderUtils.createIfMatchHeaders;
+import static com.epam.aidial.cfg.utils.HeaderUtils.createIfNonMatchHeaders;
+
+@Slf4j
+@Service
+@RequiredArgsConstructor
+@LogExecution
+public class ToolSetResourceService implements ResourceService {
+    private static final String BASE_PATH = "public/";
+
+    private final ToolSetClient toolSetClient;
+    private final ToolSetClientMapper toolSetClientMapper;
+    private final ResourceClient resourceClient;
+    private final ResourceClientMapper resourceClientMapper;
+
+    @Value("${core.toolsets.metadata.default.limit}")
+    private int toolSetsMetadataDefaultLimit;
+
+    public ToolSetResourceNodeInfo getToolSetResources(ResourceMetadataRequest request) {
+        var toolSetMetadataDto = getMetadata(request);
+        return toolSetClientMapper.toToolSetInfo(toolSetMetadataDto);
+    }
+
+    @Override
+    public FolderInfo getFolders(ResourceMetadataRequest request) {
+        try {
+            ToolSetMetadataDto toolSetMetadataDto = getMetadata(request);
+            return toolSetClientMapper.toFolderInfo(toolSetMetadataDto, TOOLSETS_PREFIX);
+        } catch (FeignException.FeignClientException.NotFound notFound) {
+            log.debug("Toolset metadata not found for request: {}", request, notFound);
+            return null;
+        }
+    }
+
+    @Override
+    public ToolSetMetadataDto getMetadata(ResourceMetadataRequest request) {
+        var recursive = request.isRecursive();
+        var nextToken = request.getNextToken();
+        var path = request.getPath() != null ? request.getPath() : BASE_PATH;
+        var limit = request.getLimit() != null ? request.getLimit() : toolSetsMetadataDefaultLimit;
+        return toolSetClient.getToolSetMetadata(path, recursive, nextToken, limit);
+    }
+
+    @Override
+    public void move(MoveResource moveResource) {
+        var moveResourceDto = resourceClientMapper.toMoveResourceDto(moveResource, TOOLSETS_PREFIX);
+        resourceClient.move(moveResourceDto);
+    }
+
+    @Override
+    public ResourceType getResourceType() {
+        return ResourceType.TOOL_SET;
+    }
+
+    public ToolSetResource getToolSetResource(String path) {
+        return fetchToolSetResource(path, null).model();
+    }
+
+    public DomainModelWithEtag<ToolSetResource> getToolSetResource(String path, String etag) {
+        return fetchToolSetResource(path, etag);
+    }
+
+    private DomainModelWithEtag<ToolSetResource> fetchToolSetResource(String path, String etag) {
+        var response = toolSetClient.getToolSetResource(path, createIfNonMatchHeaders(etag));
```

Comment:
can response be NOT_MODIFIED without an body/toolset?

Reply:
@siarhei-fedziukovich, This comment is incorrect; it referred to the behavior of the put method. Please ignore it in this context.

The correct behavior is described in the [comment](https://github.com/epam/ai-dial-admin-backend/pull/274#discussion_r2381923712)

Reply:
ok. i ask first question again. can response be NOT_MODIFIED without an body/toolset? what ToolSetResourceController returns in that case (if  toolSetClient.getToolSetResource(path, createIfNonMatchHeaders(etag)) returns null)? could you prepare a test for this case pls.